### PR TITLE
Updating reporters to all use serde for both json and yaml + misc improvements

### PR DESCRIPTION
**Pull request by joshfried-aws**
*Issue #, if available:*
#350 

*Description of changes:*
Modified the response for the generic reporter to use serde_yaml/json when the output format is yaml or json to have consistency with other reporters. 

Also made some changes that were done in #317 but never merged. Thank you @dchakrav-github. 

This introduces 2 changes:
1 - we now use serde_json's pretty print feature for json output with the parse-tree command. This makes the output easier to read. This caused a need to update tests where we spied on the printed output to console. 

2 - improve error logging 

Misc changes

- added tests for parse-tree when a function expression is present in a rule. 
- improve integration tests for tests where a function call is present

---
*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: guard-ffi/src/errors.rs [hunk 1]
```diff
@@ -29,6 +29,7 @@ fn get_code(e: &Error) -> ErrorCode {
         Error::MissingValue(_err) => 16,
         Error::FileNotFoundError(_) => 17,
         Error::IllegalArguments(_) => 18,
+        Error::InternalError(_) => unreachable!(),
```
Comment (akshayrane) [inline guard-ffi/src/errors.rs:32]: Some of the changes from PR #372 are getting duplicated here?

File: guard/resources/test-command/functions/data/template.yaml [hunk 2]
```diff
@@ -5,17 +5,27 @@
       newServer:
         Type: AWS::New::Service
         Properties:
-          Policy: |
```
Comment (akshayrane) [inline guard/resources/test-command/functions/data/template.yaml:8]: Was this template not being used elsewhere before?

File: guard/src/commands/parse_tree.rs [hunk 3]
```diff
@@ -71,7 +71,7 @@ impl Command for ParseTree {
         let rules = crate::rules::parser::rules_file(span)?;
 
         match app.get_flag(PRINT_JSON.0) {
-            true => serde_json::to_writer(writer, &rules)?,
+            true => serde_json::to_writer_pretty(writer, &rules)?,
             false => serde_yaml::to_writer(writer, &rules)?,
```
Comment (akshayrane) [inline guard/src/commands/parse_tree.rs:75]: Should we update the writer for false arm as well?

File: guard/src/commands/validate/cfn.rs [hunk 4]
```diff
@@ -81,19 +81,39 @@ impl<'reporter> Reporter for CfnAware<'reporter> {
...
+                OutputFormatType::SingleLineSummary => {
+                    match single_line(
+                        write,
+                        data_file,
+                        data_file_bytes,
+                        rules_file,
+                        data,
+                        failure_report,
+                    ) {
+                        Err(crate::Error::InternalError(_)) => {
```
Comment (akshayrane) [inline guard/src/commands/validate/cfn.rs:99]: duplicate changes again?

File: guard/src/commands/parse_tree.rs (hunk 3 above)
Comment (joshfried-aws) [reply guard/src/commands/parse_tree.rs:75]: That is not an option for serde_yaml

File: guard/resources/test-command/functions/data/template.yaml (hunk 2 above)
Comment (joshfried-aws) [reply guard/resources/test-command/functions/data/template.yaml:8]: it was being used for a unit test in test_command.rs

**Review by akshayrane — APPROVED** [review]

**Pull request merged**